File: app/src/main/java/com/ihuntto/motioneventwriter/MainActivity.java
package com.ihuntto.motioneventwriter;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;

public class MainActivity extends AppCompatActivity {
    private static final int REQUEST_ACCESS_EXTERNAL_STORAGE = 0x01;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        PointerView view = new PointerView(this);
        setContentView(view);

        ActivityCompat.requestPermissions(MainActivity.this,
                new String[]{Manifest.permission.READ_EXTERNAL_STORAGE,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE},
                REQUEST_ACCESS_EXTERNAL_STORAGE);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (grantResults.length == 0 || grantResults[0] == PackageManager.PERMISSION_DENIED) {
            if (requestCode == REQUEST_ACCESS_EXTERNAL_STORAGE) {
                Toast.makeText(this, "Need access external storage.", Toast.LENGTH_LONG).show();
            }
            finish();
        }
    }
}

File: app/src/main/java/com/ihuntto/motioneventwriter/MainActivity.kt
package com.ihuntto.motioneventwriter

import android.os.Bundle
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.result.contract.ActivityResultContracts
import com.ihuntto.motioneventwriter.databinding.ActivityMainBinding
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale


class MainActivity : ComponentActivity() {
    private lateinit var binding: ActivityMainBinding

    companion object {
        const val TAG = "MainActivity"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.getRoot())
        binding.saveBtn.setOnClickListener {
            if (!binding.pointerView.isPointersEmpty()) {
                createFileLauncher.launch(getCurrentTimeFileName("pointer_data"))
            } else {
                Toast.makeText(this, R.string.empty_pointers, Toast.LENGTH_SHORT).show()
            }
        }
    }

    private val createFileLauncher =
        registerForActivityResult(ActivityResultContracts.CreateDocument("text/plain")) { uri ->
            if (uri == null) {
                Toast.makeText(this, R.string.no_file_selected, Toast.LENGTH_SHORT).show()
            } else {
                val pointers = binding.pointerView.getPointers()
                contentResolver.openOutputStream(uri)?.let {
                    write(it, pointers)
                }
            }
        }

    fun getCurrentTimeFileName(prefix: String = "file", extension: String = "txt"): String {
        val timeStamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault()).format(Date())
        return "${prefix}_${timeStamp}.$extension"
    }
}
